### PR TITLE
`Exam mode`: Improve exam submission performance 

### PR DESCRIPTION
**Pull request by maximiliansoelch**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

We want to speed up the process of handing in the exam and avoid unnecessary database calls.

### Description

This PR checks if the text/model/quiz submitted with handing in the exam differs from the previous submission.
If yes, the final submission of the exercise is saved one final time.
If not, the final submission is **not** saved again!

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Quiz, Modeling, and Text exercise

1. Participate in the exam as a student
2. Make sure that the exam conduction works as usual
3. After handing in the exam, make sure that the exam submission summary shows the answers you actually made during the exam

#### Performance Review
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
- [ ] Code Review 3
#### Exam Mode Test
- [x] Test 1
- [x] Test 2
- [x] Test 3

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [DragAndDropSubmittedAnswer.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.quiz/DragAndDropSubmittedAnswer.html) | 91% | ✅ |
| [MultipleChoiceSubmittedAnswer.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.quiz/MultipleChoiceSubmittedAnswer.html) | 93% | ✅ |
| [ShortAnswerSubmittedAnswer.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.quiz/ShortAnswerSubmittedAnswer.html) | 87% | ✅ |
| [StudentExamRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/StudentExamRepository.html) | 95% | ✅ |
| [SubmissionVersionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/SubmissionVersionService.html) | 87% | ✅ |
| [StudentExamService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/StudentExamService.html) | 92% | ✅ |
| [StudentExamResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5194/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/StudentExamResource.html) | 92% | ✅ |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java [hunk 1]
```diff
@@ -257,10 +257,18 @@ else if (submittedAnswer instanceof ShortAnswerSubmittedAnswer) {
...
+                            TextSubmission textSubmission = (TextSubmission) submission;
+                            if (!textSubmission.getText().equals(existingSubmission.getText())) {
+                                textSubmissionRepository.save(textSubmission);
+                            }
                         }
                         else if (exercise instanceof ModelingExercise) {
-                            modelingSubmissionRepository.save((ModelingSubmission) submission);
+                            ModelingSubmission existingSubmission = (ModelingSubmission) existingParticipation.findLatestSubmission().get();
+                            ModelingSubmission modelingSubmission = (ModelingSubmission) submission;
+                            if (!modelingSubmission.getModel().equals(existingSubmission.getModel())) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java:269]: Is `getModel()` nullable (e.g. in case of empty submissions) ?

Comment (maximiliansoelch) [reply src/main/java/de/tum/in/www1/artemis/service/exam/StudentExamService.java:269]: Yes, it is. Thanks for pointing it out!

**Review by RY997 — APPROVED** [review]
Tested on ts5 in testing session, worked well.

**Review by pal03377 — APPROVED** [review]
Tested on TS5 (exam testing session). We tested it with a text exercise, a modeling exercise and a quiz exercise and each changed something last second (not saved). It still was visible in the student summary and for the text and modeling exercise also in the assessment view for the instructor (we could not find a view for quizzes). ✅

**Review by FelixTJDietrich — APPROVED** [review]
Tested on ts5 during exam testing session, works as expected.

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts5. My submission was still saved when I changed something in the last 30s

**Review by basak-akan — APPROVED** [review]
Tested on ts5 during exam testing session, worked as expected. Last-second changes successfully saved.

File: src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java [hunk 2]
```diff
@@ -117,6 +118,11 @@ public <T extends Exercise> T getFirstExerciseWithType(Exam exam, Class<T> clazz
         return (T) exercise;
     }
 
+    public <T extends Exercise> T getFirstExerciseWithType(StudentExam studentExam, Class<T> clazz) {
+        var exercise = studentExam.getExercises().stream().filter(ex -> ex.getClass().equals(clazz)).findFirst().orElseThrow();
```
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/exercise/ExerciseUtilService.java:122]:
```suggestion
        var exercise = studentExam.getExercises().stream().filter(clazz::isInstance).findFirst().orElseThrow();
```

**Review by MaximilianAnzinger — COMMENTED** [review]
Left a comment. Other than that code LGTM 👍

**Review by krusche — APPROVED** [review]

**Pull request merged**